hera_cal/smooth_cal.py

```diff
@@ -105,9 +106,109 @@ def time_filter(gains, wgts, times, filter_scale=1800.0, nMirrors=0):
+    fringe_scale = (time_scale)**-1  # in Hz
+
+    # find per-integration delays, smooth on the time_scale of gain smoothing, and rephase 
+    taus = fft_dly(gains, df, wgts, medfilt=False, solve_phase=False)[0].astype(np.complex)  # delays are in seconds
+    if not np.all(taus == 0):  # this breaks CLEAN, but it means we don't need smoothing anyway
+        taus = uvtools.dspec.high_pass_fourier_filter(taus.T, np.sum(wgts, axis=1, keepdims=True).T,
+                                                      fringe_scale, dt, tol=tol, maxiter=maxiter)[0].T
+    rephasor = np.exp(-2.0j * np.pi * np.outer(np.abs(taus), freqs))
+    
+    # Build fourier space image and kernel for deconvolution
+    window = aipy.dsp.gen_window(len(freqs), window=window, **win_kwargs)
+    image = np.fft.ifft2(gains * rephasor * wgts * window)
+    kernel = np.fft.ifft2(wgts * window)
+    
+    # set up "area", the set of Fourier modes that are allowed to be non-zero in the CLEAN
+    if filter_mode == 'rect':
+        area = np.outer(np.where(np.abs(fringes) < fringe_scale, 1, 0), 
+                        np.where(np.abs(delays) < delay_scale, 1, 0))
+    elif filter_mode == 'plus':
+        area = np.zeros(image.shape, dtype=int)
+        area[0] = np.where(np.abs(delays) < delay_scale, 1, 0)
+        area[:, 0] = np.where(np.abs(fringes) < fringe_scale, 1, 0)
+    else:
+        raise ValueError("CLEAN mode {} not recognized. Must be 'rect' or 'plus'.".format(filter_mode))
+
+    # perform deconvolution
+    CLEAN, info = aipy.deconv.clean(image, kernel, tol=tol, area=area, stop_if_div=False, maxiter=maxiter)
+    filtered = np.fft.fft2(CLEAN + info['res'] * area)
+    del info['res']  # this matches the convention in uvtools.dspec.high_pass_fourier_filter
+    return filtered / rephasor, info
+
+
+def pick_reference_antenna(flags):
+    '''Pick a refrence antenna that has the minimum number of per-antenna flags.
+    
+    Arguments:
+        flags: dictionary mapping antenna keys to flag waterfall where True means flagged
+
+    Returns:
+        refant: key of the antenna with the minimum number of flags. Tie goes to the first in a sorted list of keys.
+    '''
+    flags_per_ant = {ant: np.sum(f) for ant, f in flags.items()}
+    refant = sorted([ant for ant, nflags in flags_per_ant.items() if nflags == np.min(flags_per_ant.values())])[0]
+    return refant
+
+
+def rephase_to_refant(gains, refant, flags=None):
+    '''Rephase all gains in place so that the phase of the reference antenna is 0 for all times and frequencies.
+
+    Arguments:
+        gains: Dictionary mapping antenna keys to gain waterfalls. Modified in place.
+        refant: Antenna key of antenna to make the reference antenna
+        flags: Optional dictionary mapping antenna keys to flag waterfall.
+            Used only to verify that all gains are flagged where the refant is flagged.
+    '''
+    refant_phasor = gains[refant] / np.abs(gains[refant])
```

Comment:
Is it possible you will be dividing by zero here (e.g. in the case of flagged data)? You may want to avoid this with true_divide so you don't get blasted with warnings.

Reply:
AFAIK, we'll be putting 1.0s rather than 0s for flagged gains. But this is a true divide, since we're importing division from __future__

If we get lots of warnings, we can later decide to suppress or deal with them.

Reply:
Ok, if we don't expect zero gains in a regular mode, then getting warnings is probably good.